src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+                            }
+
+                            break;
+                        case SyntaxKind.OrderByClause:// OrderBy is not supported by foreach.
+                        default:
+                            return false;
+                    }
+                }
+
+                var selectOrGroupClause = queryBody.SelectOrGroup;
+                switch (selectOrGroupClause.Kind())
+                {
+                    case SyntaxKind.SelectClause:
+                        // TODO add a unit test with multiple selects
+                        var selectClause = (SelectClauseSyntax)selectOrGroupClause;
+
+                        // End of the queue
+                        if (queryBody.Continuation == null)
+                        {
+                            processingState.AddStatementProcessingMethod(s => WrapWithBlock(leafExpressionCreationMethod(((SelectClauseSyntax)selectOrGroupClause).Expression)));
+                        }
+                        else
+                        {
+                            processingState.AddVariableName(queryBody.Continuation.Identifier.ValueText);
+                            processingState.AddStatementProcessingMethod(
+                                s => AddToBlockTop(
+                                     SyntaxFactory.LocalDeclarationStatement(
+                                        SyntaxFactory.VariableDeclaration(
+                                            VarNameIdentifier,
+                                            SyntaxFactory.SingletonSeparatedList(
+                                                SyntaxFactory.VariableDeclarator(
+                                                    queryBody.Continuation.Identifier,
+                                                    null,
+                                                    SyntaxFactory.EqualsValueClause(selectClause.Expression))))), s));
+
+                            TryProcessQueryBody(leafExpressionCreationMethod, queryBody.Continuation.Body, processingState);
+                        }
+                        break;
+                    case SyntaxKind.GroupClause:
+                        // Groupclause is not supported by the conversion
+                        return false;
+                    default:
+                        throw new ArgumentException(selectOrGroupClause.Kind().ToString());
+                }
+
+                return true;
+            }
+
+            private static BlockSyntax AddToBlockTop(StatementSyntax newStatement, StatementSyntax statement)
+            {
+                if (statement.Kind() == SyntaxKind.Block)
+                {
+                    var list = new List<StatementSyntax>();
+                    list.Add(newStatement);
+                    list.AddRange(((BlockSyntax)statement).Statements); // TODO refactor
+                    return SyntaxFactory.Block(list);
```

Comment:
this will throw away the curlies from the original block.  PLease keep them.

Reply:
Those are blocks are generating by me. For example, I could already generate 
foreach(...) 
{
}

Now I would like to add 
var a = b;
at the top of this foreach. I know that it should be wrapped with a block for further usage.
{
var a = b;
foreach(...) 
{
}
}

Now, I found that I need to add one more declaration about like 
var c = d;
I would prefer to join them together like:
{
var c = d;
var a = b;
foreach(...) 
{
}
}
rather than 
{
var c = d;
{
var a = b;
foreach(...) 
{
}
}
}

---
In reply to: [175267941](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267941) [](ancestors = 175267941)

Reply:
That's fine.  My only point is that you already have a block, so it seems cleaner and nicer to just do ```((BlockSyntax)statement).WithStatements(...)``` rather than SyntaxFactory.Block

Reply:
Made some improvements there however
WithStatements requires SyntaxList as an argument where SyntaxFactory.Block allows IEnumerable.

---
In reply to: [175617388](https://github.com/dotnet/roslyn/pull/25362#discussion_r175617388) [](ancestors = 175617388)